obspy/clients/fdsn/mass_downloader/download_helpers.py

```diff
@@ -0,0 +1,1140 @@
+        Should be run after the MiniSEED and StationXML downloads finished.
+        It will make sure that every MiniSEED file also has a corresponding
+        StationXML file.
+        """
+        for station in self.stations.values():
+            station.sanitize_downloads(logger=self.logger)
+
+    def _check_downloaded_data(self):
+        """
+        Read the downloaded data, set the proper status flags and a remove
+        data that does not meet the QC criteria. It just checks the
+        downloaded data for minimum length and gaps/overlaps.
+
+        Returns the downloaded_bytes and the discarded_bytes.
+        """
+        downloaded_bytes = 0
+        discarded_bytes = 0
+        for sta in self.stations.values():
+            for cha in sta.channels:
+                for interval in cha.intervals:
+                    # The status of the interval should not have changed if
+                    # it did not require doenloading in the first place.
+                    if interval.status != STATUS.NEEDS_DOWNLOADING:
+                        continue
+
+                    # If the file does not exist, mark the time interval as
+                    # download failed.
+                    if not os.path.exists(interval.filename):
+                        interval.status = STATUS.DOWNLOAD_FAILED
+                        continue
+
+                    size = os.path.getsize(interval.filename)
+                    if size == 0:
+                        self.logger.warning("Zero byte file '%s'. Will be "
+                                            "deleted." % interval.filename)
+                        utils.safe_delete(interval.filename)
+                        interval.status = STATUS.DOWNLOAD_FAILED
+                        continue
+
+                    # Guard against faulty files.
+                    try:
+                        st = obspy.read(interval.filename, headonly=True)
+                    except Exception as e:
+                        self.logger.warning(
+                            "Could not read file '%s' due to: %s\n"
+                            "Will be discarded." % (interval.filename, str(e)))
+                        utils.safe_delete(interval.filename)
+                        discarded_bytes += size
+                        interval.status = STATUS.DOWNLOAD_FAILED
+                        continue
+
+                    # Valid files with no data.
+                    if len(st) == 0:
+                        self.logger.warning(
+                            "Empty file '%s'. Will be deleted." %
+                            interval.filename)
+                        utils.safe_delete(interval.filename)
+                        discarded_bytes += size
+                        interval.status = STATUS.DOWNLOAD_FAILED
+                        continue
+
+                    # If user did not want gappy files, remove them.
+                    if self.restrictions.reject_channels_with_gaps is True and\
```

Comment:
Yea I know - I still like to do this as Python's dynamic typing has bitten me too many times...